### PR TITLE
Update `--profile` flag to override default chain's profile rather than replacing default chain

### LINKED ISSUE #389: `--profile` does not correctly use `credential_process` (Appstream)
### Mountpoint for Amazon S3 version

mountpoint-s3 0.3.0-d0ef0b9

### AWS Region

us-west-2

### Describe the running environment

Trying to mount S3 bucket in Appstream using IAM [Profile](https://docs.aws.amazon.com/appstream2/latest/developerguide/using-iam-roles-to-grant-permissions-to-applications-scripts-streaming-instances.html#how-to-use-iam-role-with-streaming-instances)

I get an error.

Command:
mount-s3 --read-only --allow-other --profile appstream_machine_role --region us-west-2 -f BUCKET_NAME /mnt/ 

### What happened?

Error:

Caused by:
    0: HeadBucket failed for bucket BUCKET_NAME in region us-west-2
    1: Client error
    2: Unknown response error: MetaRequestResult { response_status: 0, crt_error: Error(6146, "aws-c-auth: AWS_AUTH_SIGNING_NO_CREDENTIALS, Attempt to sign an http request without credentials"), error_response_headers: None, error_response_body: None }

### Relevant log output

```shell
2023-07-18T18:37:05.280819Z DEBUG awscrt::AWSProfile: Creating profile collection from file at "/home/ImageBuilderAdmin/.aws/config"    
2023-07-18T18:37:05.280877Z TRACE awscrt::AWSProfile: Parsing aws profile line in profile "<None>", current property: "<None>"    
2023-07-18T18:37:05.280927Z TRACE awscrt::AWSProfile: Parsing aws profile line in profile "appstream_machine_role", current property: "<None>"    
2023-07-18T18:37:05.280963Z DEBUG awscrt::AuthCredentialsProvider: (id=0x561d1bb35cd0) Profile credentials provider successfully built config profile collection from file at (/home/ImageBuilderAdmin/.aws/config)    
2023-07-18T18:37:05.280994Z DEBUG awscrt::AWSProfile: Creating profile collection from file at "/home/ImageBuilderAdmin/.aws/credentials"    
2023-07-18T18:37:05.281042Z TRACE awscrt::AWSProfile: Parsing aws profile line in profile "<None>", current property: "<None>"    
2023-07-18T18:37:05.281077Z  WARN awscrt::AWSProfile: Profile declarations in credentials files are not allowed to begin with the "profile" keyword    
2023-07-18T18:37:05.281109Z  WARN awscrt::AWSProfile: Profile Parse context:
 Source File:/home/ImageBuilderAdmin/.aws/credentials
 Line: 1
 Current Profile: <None>
 Current Property: <None>    
2023-07-18T18:37:05.281140Z TRACE awscrt::AWSProfile: Parsing aws profile line in profile "<None>", current property: "<None>"    
2023-07-18T18:37:05.281171Z  WARN awscrt::AWSProfile: Property definition seen outside a profile    
2023-07-18T18:37:05.281200Z  WARN awscrt::AWSProfile: Profile Parse context:
 Source File:/home/ImageBuilderAdmin/.aws/credentials
 Line: 2
 Current Profile: <None>
 Current Property: <None>    
2023-07-18T18:37:05.281233Z DEBUG awscrt::AuthCredentialsProvider: (id=0x561d1bb35cd0) Profile credentials provider successfully built credentials profile collection from file at (/home/ImageBuilderAdmin/.aws/credentials)    
2023-07-18T18:37:05.281268Z  WARN awscrt::AWSProfile: property "credential_process" has value "" replaced during merge    
2023-07-18T18:37:05.281301Z  INFO awscrt::AuthCredentialsProvider: (id=0x561d1bb35cd0) Profile credentials provider attempting to pull credentials from profile "appstream_machine_role"    
2023-07-18T18:37:05.281335Z ERROR awscrt::AuthSigning: (id=0x7f9fc4000ba0) Credentials Provider failed to source credentials with error 2049(aws-c-http: AWS_ERROR_HTTP_HEADER_NOT_FOUND, The specified header was not found)    
2023-07-18T18:37:05.281368Z ERROR awscrt::S3MetaRequest: id=0x561d1ba71d20 Meta request could not sign HTTP request due to error code 6146 (Attempt to sign an http request without credentials)    
2023-07-18T18:37:05.281399Z ERROR awscrt::S3MetaRequest: id=0x561d1ba71d20 Could not prepare request 0x7f9fcc001790 due to error 6146 (Attempt to sign an http request without credentials).
```

### PR DESCRIPTION
**Pull request by dannycjones**
## Description of change

_A previous PR was open for this change but was very stale: #415._

This change means that a profile can be specified that does not configure any credentials, but may configure other things. **Mountpoint does not read any configuration from an AWS Profile other than credentials today** but may do in the future.

Before this change, the entire credential chain was replaced with one that would supply profile credentials or error - no fallback. This was not in line with AWS CLI or SDKs.

This should fix #389 since the default chain also allows credential process credentials to be used.

## Does this change impact existing behavior?

Yes, it fixes the behavior of the `--profile <AWS_PROFILE>` flag to be in line with `AWS_PROFILE` environment variable, as well as `--profile <AWS_PROFILE>` flag of the AWS CLI.

It also fixes bug #389.

**Note:** It does not validate that the AWS profile exists, like the AWS CLI does. If the profile does not exist, the parameter will be ignored. Mountpoint already behaves this way today when using the `AWS_PROFILE` environment variable. Mountpoint previously reporting an error was a side effect of replacing the default chain with a profile provider.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

**Comment by dannycjones** [comment]
Opened #685 to address crate builds issue.

File: mountpoint-s3-client/src/s3_crt_client.rs [hunk 1]
```diff
@@ -159,19 +157,27 @@ impl S3ClientConfig {
...
 /// Authentication configuration for the CRT-based S3 client
-#[derive(Debug, Clone, Default)]
+#[derive(Clone, Debug)]
 pub enum S3ClientAuthConfig {
     /// The default AWS credentials resolution chain, similar to the AWS CLI
-    #[default]
-    Default,
+    DefaultChain {
+        /// Optional profile override to be used when evaluating credential chain
+        profile_name_override: Option<String>,
```
Comment (jamesbornholt) [inline mountpoint-s3-client/src/s3_crt_client.rs:165]: naming nit: I think `Option` already captures that it's an override of sorts, so maybe just `profile_name`

File: mountpoint-s3/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,10 @@
 ## Unreleased
 
+### Other changes
+* Option `--profile <AWS_PROFILE>` no longer replaces the entire default credential chain.
```
Comment (jamesbornholt) [inline mountpoint-s3/CHANGELOG.md:4]: This is describing the implementation, not the customer-visible behavior. I'd just mention the issue we're fixing (credential process not working). Also, this is a breaking change -- we no longer validate incorrect profile names, and (if my other comment is right) also now respect env vars when profile is specified.

**Review by jamesbornholt — COMMENTED** [review]
What's the CLI behavior if you both specify `--profile` and use environment variables to provide creds? It looks like the CRT default chain [checks env vars first](https://github.com/awslabs/aws-c-auth/blob/main/source/credentials_provider_default_chain.c#L316), so this is technically a breaking change for Mountpoint (env var credentials weren't respected with `--profile` before but now are). Knowing what the CLI does would be useful as a baseline.

File: mountpoint-s3/CHANGELOG.md (hunk 2 above)
Comment (dannycjones) [reply mountpoint-s3/CHANGELOG.md:4]: You're absolutely right with the environment variables, that's why I ended up purging them from the test environment. Let me clarify that in the change log.

I'll also test the AWS CLI since I think you may be right about `--profile` being different from `AWS_PROFILE` in the CLI.

**Comment by dannycjones** [comment]
TLDR: Current implementation in this PR is wrong and making a bad change to `--profile` which should not use environment credentials.

### Reviewing `--profile` and `AWS_PROFILE` behaviors

This compares the AWS CLI and other SDKs. I configured credentials in the environment (`AWS_ACCESS_KEY_ID`, etc.) and some credentials for the given profile. I then configured the profile using either the environment variable or CLI argument (if applicable).

| Tool | `AWS_PROFILE` | `--profile` |
|--------|--------|--------|
| AWS CLI | Environment creds | Profile creds |
| Mountpoint v1.3.1 | Environment creds | Profile creds (broken on cred process ⚠️) |
| Mountpoint (THIS CHANGE) | Environment creds | Environment creds ⚠️ |
| Rust SDK | [Environment creds](https://github.com/awslabs/aws-sdk-rust/blob/main/sdk/aws-config/src/default_provider/credentials.rs#L27-L32) | N/A |
| Java SDK | [Environment creds](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.html) | N/A |

I was misunderstanding `--profile` and it does ignore the environment variable and is different from `AWS_PROFILE`.


Here's some further testing where I configured credentials in the environment and varied the contents of the profile to be super clear about the behavior.

| Profile | AWS CLI w/ `AWS_PROFILE` | AWS CLI w/ `--profile` |
|--------|--------|--------|
| Exists w/ creds | Environment creds | Profile creds |
| Exists w/o creds | Environment creds | "Unable to locate credentials." |
| Doesn't exist | "The config profile (\<PROFILE\>) could not be found" | "Unable to locate credentials." |

Ultimately, I think I need to ensure the `--profile` behavior doesn't change, and that we go and ensure that the credential process is fixed (which was the original bug). Thanks for flagging this, James!

**Comment by dannycjones** [comment]
I've updated this PR by adding tests about the issue discovered, but it'll still need some changes in the AWS CRT so that we can fix the credential chain for `--profile` before being ready for merge/review.

File: mountpoint-s3/CHANGELOG.md (hunk 2 above)
Comment (dannycjones) [reply mountpoint-s3/CHANGELOG.md:4]: CRT folks have added an option to disable ENV credentials on the default chain, so we can pick that up to avoid changing the behavior there. https://github.com/awslabs/aws-c-auth/commit/3d9f13fd6610d2c9f5ef5cf404bce4d9aeaca15b

**Comment by dannycjones** [comment]
Closing this out since:

- The code base has changed a lot since the original change.
- The original issue (https://github.com/awslabs/mountpoint-s3/issues/389) is now resolved as the CRT now correctly includes the credential process in profile providers.

We should still consider replacing the use of the profile credential provider in Mountpoint with the default chain and a profile override, however I would not currently prioritize this due to a) this issue not being reported and b) the possibility that we remove Mountpoint's dependency on `mountpoint-s3-client` in the foreseeable future.
